python/cudf/cudf/utils/dtypes.py

```diff
@@ -494,6 +494,15 @@ def find_common_type(dtypes):
     if len(dtypes) == 0:
         return None
 
+    # Early exit for categoricals since they're not hashable and therefore
+    # can't be put in a set.
+    if any(is_categorical_dtype(dtype) for dtype in dtypes):
+        # TODO: Is there a better behavior here? We could check if two distinct
+        # dtype objects actually had the same categories. Alternatively, should
+        # we actually combine categories (that seems wrong, but is also
+        # possible).
+        return cudf.dtype("O")
+
```

Comment:
I agree about it being nice to put a better behavior here, is there a precedent for that?

Reply:
I don't think so, perhaps @isVoid would know if there's a better choice to make here?

Reply:
To me, there are two folds to this question. The first is thinking about subtyping of categorical types can extend far into set theories, for example:

column a that has categorical dtype `[a, b, c]` and column b has categorical dtype [a, c], then many (all?) ops that's applicable to column a is also applicable to column b. So perhaps the common type is the intersection of all categorical dtypes?

The other fold is, I'm curious why this function is even looped in here for this refactor? `find_common_types` was meant to be just a shim layer over np.find_common_type to handle some special corner cases - that should just deal with np dtypes. Were some other types also gets passed in here too?

Reply:
I agree with the two points you make there. Thinking it for a second time, your proposal using the union of the categories does make sense. The common type of two numerical type is the one that's sufficiently large enough, that when being casted to, every possible number maintain their original value. For the categories, the union of all categories has the same property.